compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -599,6 +599,22 @@ object Objects:
 
           case _ => a
 
+    def filterType(tpe: Type)(using Context): Value =
+      val baseClasses = tpe.baseClasses
+      if baseClasses.isEmpty then a
+      else filterClass(baseClasses.head)
+
+    def filterClass(sym: Symbol)(using Context): Value =
+        if !sym.isClass then a
+        else
+          val klass = sym.asClass
+          a match
+            case Cold => Cold
+            case ref: Ref if ref.klass.isSubClass(klass) => ref
+            case ref: Ref => Bottom
+            case ValueSet(values) => values.map(v => v.filterClass(klass)).join
+            case _ => a // TODO: could be more precise for OfArray; possibly add class information for Fun
```

Comment:
As discussed in the meeting, we can further filter `OfArray` here. Meanwhile, we can remove the old code that partially handles this.